tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp

```diff
@@ -287,6 +287,92 @@ void test_two_index_constraint_analytic(
       make_with_value<decltype(two_index_constraint)>(x, 0.0),
       numerical_approx);
 }
+
+// Test the return-by-value four-index constraint function using random values
+template <size_t SpatialDim, typename Frame, typename DataType>
+void test_four_index_constraint_random(const DataType& used_for_size) noexcept {
+  pypp::check_with_random_values<1>(
+      static_cast<tnsr::iaa<DataType, SpatialDim, Frame> (*)(
+          const tnsr::ijaa<DataType, SpatialDim, Frame>&)>(
+          &GeneralizedHarmonic::four_index_constraint<SpatialDim, Frame,
+                                                      DataType>),
+      "TestFunctions", "four_index_constraint", {{{-10.0, 10.0}}},
+      used_for_size);
+}
+
+// Test the return-by-reference four-index constraint
+// by comparing to a time-independent analytic solution (e.g. Kerr-Schild)
+template <typename Solution>
+void test_four_index_constraint_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound,
+    const double error_tolerance) noexcept {
+  // Shorter names for tags.
+  using Phi = ::GeneralizedHarmonic::Phi<3, Frame::Inertial>;
+  using VariablesTags = tmpl::list<Phi>;
+
+  // Check vs. time-independent analytic solution
+  // Set up grid
+  const size_t data_size = pow<3>(grid_size_each_dimension);
+  Mesh<3> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+
+  using Affine = CoordinateMaps::Affine;
+  using Affine3D = CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1.0, 1.0, lower_bound[0], upper_bound[0]},
+          Affine{-1.0, 1.0, lower_bound[1], upper_bound[1]},
+          Affine{-1.0, 1.0, lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = std::numeric_limits<double>::signaling_NaN();
+
+  // Evaluate analytic solution
+  const auto vars =
+      solution.variables(x, t, typename Solution::template tags<DataVector>{});
+  const auto& lapse = get<gr::Tags::Lapse<>>(vars);
+  const auto& d_lapse =
+      get<typename Solution::template DerivLapse<DataVector>>(vars);
+  const auto& shift = get<gr::Tags::Shift<3>>(vars);
+  const auto& d_shift =
+      get<typename Solution::template DerivShift<DataVector>>(vars);
+  const auto& spatial_metric = get<gr::Tags::SpatialMetric<3>>(vars);
+  const auto& d_spatial_metric =
+      get<typename Solution::template DerivSpatialMetric<DataVector>>(vars);
+
+  // Compute derivatives d_psi, d_phi, d_pi, and d_gauge_function numerically
+  Variables<VariablesTags> gh_vars(data_size);
+  auto& phi = get<Phi>(gh_vars);
```

Comment:
phi isn't used below, so the intermediate value isn't needed, just the values set in the Variables